src/client/driver/index.js

```diff
@@ -0,0 +1,105 @@
+import hammerhead from './deps/hammerhead';
+import testCafeCore from './deps/testcafe-core';
+import testCafeUI from './deps/testcafe-ui';
+
+import MESSAGE from '../../test-run/client-messages';
+import COMMAND_TYPE from '../../test-run/commands/type';
+import { UncaughtJSError } from '../../errors/test-run';
+
+import * as browser from '../browser';
+import executeActionCommand from './command-executors/execute-action-command';
+import ContextStorage from './storage';
+
+var transport         = hammerhead.transport;
+var XhrBarrier        = testCafeCore.XhrBarrier;
+var pageUnloadBarrier = testCafeCore.pageUnloadBarrier;
+var eventUtils        = testCafeCore.eventUtils;
+var modalBackground   = testCafeUI.modalBackground;
+
+
+const executingCommandFlag = 'testcafe|driver|executing-command-flag';
+
+
+export default class ClientDriver {
+    constructor (testRunId, heartbeatUrl, browserStatusUrl) {
+        this.testRunId        = testRunId;
+        this.heartbeatUrl     = heartbeatUrl;
+        this.browserStatusUrl = browserStatusUrl;
+        this.contextStorage   = new ContextStorage(window, testRunId);
+
+        this.pageInitialXhrBarrier = new XhrBarrier();
+
+        pageUnloadBarrier.init();
+    }
+
+    start () {
+        browser.startHeartbeat(this.heartbeatUrl, hammerhead.nativeMethods.XMLHttpRequest);
+
+        modalBackground.initAndShowLoadingText();
+        hammerhead.on(hammerhead.EVENTS.uncaughtJsError, err => this._onJsError(err));
+
+        eventUtils
+            .documentReady()
+            .then(() => this.pageInitialXhrBarrier.wait(true))
+            .then(() => {
+                var inExecuting   = this.contextStorage.getItem(executingCommandFlag);
+                var commandResult = inExecuting ? { failed: false } : null; //TODO: store errors between page reloads
+
+                modalBackground.hide();
+                this._onReady(commandResult);
+            });
+    }
+
+    _onJsError (err) {
+        return transport
+            .asyncServiceMsg({
+                cmd: MESSAGE.jsError,
+                err: new UncaughtJSError(err.msg || err.message, err.pageUrl)
+            });
+    }
+
+    _onReady (commandResult) {
+        transport
+            .asyncServiceMsg({ cmd: MESSAGE.ready, commandResult })
+            .then(command => {
+                if (command)
+                    this._onCommand(command);
+                else {
+                    //TODO:
```

Comment:
we should resend `ready` message with some interval here. But we will able to test this when we have `wait` command on the server or api v2